lib/src/repo_path.rs

```diff
@@ -83,6 +83,24 @@ impl RepoPath {
         }
     }
 
+    /// Converts workspace-relative `Path` to `RepoPath`.
+    ///
+    /// The input path should not contain `.` or `..`.
+    pub fn from_relative_path(relative_path: impl AsRef<Path>) -> Result<Self, FsPathParseError> {
+        let relative_path = relative_path.as_ref();
+        let components = relative_path
+            .components()
+            .map(|c| match c {
+                Component::Normal(a) => {
+                    let s = a.to_str().ok_or(FsPathParseError::InvalidUtf8)?;
+                    Ok(RepoPathComponent::from(s))
+                }
+                _ => Err(FsPathParseError::NotInRepo),
```

Comment:
That's correct. This function expects a normalized `Path`.

Do you think better to add a separate error type? I agree "invalid path component" is more correct here, but it's less obvious to users. It doesn't make sense to add an error type that contains both `NotInRepo` and `InvalidPathComponent`.

Reply:
The best error message I could think of was `Invalid component '..' in path`. I don't think "Invalid path" is specific enough, and I think "not in repo" is likely to confuse someone when they try to do something reasonable, like `./foo`, so, in the end, I would just have the `InvalidPathComponent` error case. Let me know what you think.

Reply:
For user-facing errors (which would be reported by `parse_fs_path()`), I think "not in the repo" is preferred than "invalid component" because ".." is valid as a file path.

".." and "." are invalid input to this function (which does literal conversion from `Path` to `RepoPath`), but they aren't to `parse_fs_path()`.

Reply:
Okay, I'll leave it to your discretion.